drivers/modem/ublox-sara-n310.c

```diff
@@ -0,0 +1,526 @@
+						  &mdata.context.iface);
+
+		k_yield();
+	}
+}
+
+/* send command */
+void n310_send(uint8_t *payload)
+{
+	if (has_ip()) {
+		int ret = 0;
+
+		LOG_INF("\033[0;37mSending command: %s", payload);
+		ret = modem_cmd_send(&mdata.context.iface,
+				     &mdata.context.cmd_handler, NULL, 0U,
+				     payload, &mdata.sem_response,
+				     K_SECONDS(1));
+
+		if (ret < 0) {
+			LOG_ERR("send function failed");
+		}
+	}
+}
+
+/* pin init */
+static int pin_init(void)
+{
+	/* set and enable pins/pwr, blank for now */
+	return 1;
+}
+
+/* create socket */
+int create_socket(int type, int port)
+{
+	int ret = 0;
+	char buf[sizeof("AT+USOCR=##,#####\r")];
+
+	snprintk(buf, sizeof(buf), "AT+USOCR=%d,%d", type, port);
+
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, buf, &mdata.sem_response, K_SECONDS(4));
+
+	modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler, NULL,
+		       0, "AT+CGPADDR=", &mdata.sem_response, K_SECONDS(2));
+
+	return ret;
+}
+
+/* close socket */
+static int close_socket(void)
+{
+	int ret = 0;
+
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, "AT+USOCL=1", &mdata.sem_response,
+			     K_SECONDS(4));
+
+	return ret;
+}
+
+/* reset module */
+void n310_modem_reset(void)
+{
+	int ret = 0, counter = 0;
+
+	pin_init();
+
+	at.CCID = "AT+CCID";
+	at.USORF = "AT+USORF=1";
+	at.NSOSTF = "AT+NSOSTF=1,\"192.168.0.20\",4242,2,4,\"0102AABB\"";
+
+	LOG_INF("Starting modem...");
+
+	static const struct setup_cmd setup_cmds[] = {
+		/* turn off echo */
+		SETUP_CMD_NOHANDLE("ATE0"),
+		/* stop functionality */
+		SETUP_CMD_NOHANDLE("AT+CFUN=0"),
+		/* extended error numbers */
+		SETUP_CMD_NOHANDLE("AT+CMEE=1"),
+		/* UNC messages for registration */
+		SETUP_CMD_NOHANDLE("AT+CREG=1"),
+		/* enable PDP context */
+		SETUP_CMD_NOHANDLE("AT+CIPCA=1"),
+		/* enable PSM */
+		SETUP_CMD_NOHANDLE("AT+CPSMS=1,,,\"01011111\",\"00000000\""),
+		/* enable HEX mode for +USOWR, +USOST, +USORD and +USORF AT commands. */
+		SETUP_CMD_NOHANDLE("AT+UDCONF=1,1"),
+
+		/* get and store modem info */
+		SETUP_CMD("AT+CGMI", "", on_cmd_atcmdinfo_manufacturer, 0U, ""),
+		SETUP_CMD("AT+CGMM", "", on_cmd_atcmdinfo_model, 0U, ""),
+		SETUP_CMD("AT+CGMR", "", on_cmd_atcmdinfo_revision, 0U, ""),
+		SETUP_CMD("AT+CGSN", "", on_cmd_atcmdinfo_imei, 0U, ""),
+
+		/* enable functionality */
+		SETUP_CMD_NOHANDLE("AT+CFUN=1"),
+	};
+
+	/* reset module */
+	ret = modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler,
+			     NULL, 0, "AT+CFUN=16", &mdata.sem_response,
+			     K_SECONDS(5));
+
+	if (ret < 0) {
+		LOG_ERR("module reset failed: %d", ret);
+		return;
+	}
+
+	modem_cmd_send(&mdata.context.iface, &mdata.context.cmd_handler, NULL,
+		       0, "AT+CCID", &mdata.sem_response, K_SECONDS(1));
+
+	/* give modem time to start responding after restart */
+	ret = -1;
+
+	while (counter++ < 50 && ret < 0) {
```

Comment:
I'm not sure if it is allowed in Zephyr to increment inside brackets. See https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_13_03.c

Reply:
It is done in a few modem drivers I looked at, so it should be fine.